### PR TITLE
Django 4.0 support

### PR DESCRIPTION
**Pull request by gasman**
Add Django 4.0 support, along with:

* dropping Django 3.0 and 3.1 support
* upgrading django-taggit to 2.x; however, django-taggit has not yet released a Django 4.0 compatible version and so this PR currently has it pinned to git master
* adding an exception for the WAGTAILFORMS_HELP_TEXT_ALLOW_HTML test for Django 4.0 and 4.0.1, which is broken on these releases due to https://code.djangoproject.com/ticket/33419 - this fix has now been merged and should be passing on the stable/4.0.x branch.

### CODE REVIEW

**Review by zerolab — APPROVED** [review]
Love a good tidy up!

Left a note and a question relating to release notes/upgrade considerations.

As an aside, I wish GitHub would show allowed failures differently The test failures here are on Django main, relating to

```
wagtail/snippets/tests.py:878: RemovedInDjango41Warning: In Django 4.1, repr() will not be called automatically on a queryset when compared to string values. Set an explicit 'transform' to silence this warning.
  self.assertQuerysetEqual(instances, ["<Advert: Test hook>"])
```

as far as I can tell. That can be a follow up task

File: docs/releases/2.16.md [hunk 1]
```diff
@@ -8,6 +8,11 @@
 
 ## What's new
 
+Django 4.0 support
+~~~~~~~~~~~~~~~~~~
+
+This release adds support for Django 4.0.
```
[the remark this answers is not in the document]
Comment (zerolab) [reply docs/releases/2.16.md:14]: Sounds like a plan

File: wagtail/admin/views/generic/models.py [hunk 2]
```diff
@@ -15,6 +20,39 @@
...
 
+if DJANGO_VERSION >= (4, 0):
+    BaseDeleteView = DjangoBaseDeleteView
+else:
+    # As of Django 4.0 BaseDeleteView has switched to a new implementation based on FormMixin
+    # where custom deletion logic now lives in form_valid:
+    # https://docs.djangoproject.com/en/4.0/releases/4.0/#deleteview-changes
+    # Here we define BaseDeleteView to match the Django 4.0 implementation to keep it consistent
+    # across all versions.
+    class BaseDeleteView(DeletionMixin, FormMixin, BaseDetailView):
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/admin/views/generic/models.py:31]: wagtail.admin.views.generic is not (yet!) a public documented part of the Wagtail extension API, so a release note isn't strictly needed - but yeah, it wouldn't hurt :-) Will add a note.

**Comment by gasman** [comment]
Force-pushed with the DeleteView release note added, and the CI matrix / dependency line commits shuffled around a bit so that we can cleanly omit the last two commits and have something that's mergeable today as an interim version - will open that as a new PR in a moment.

**Comment by gasman** [comment]
django-taggit 2.1 is now released - have rebased this without the version pinning.

**Pull request merged**